### ISSUE TITLE
Temporal Fuzzing I: Memory Models

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11176]
Temporal Fuzzing I: Memory Models<br>
<br>
https://ift.tt/BVftxmu<br>
<br>
<br>
<br>
<div><article><h4>Concurrency</h4>
<p>Reasoning about concurrent execution is hard.</p>
<p>Let’s start with some pseudocode. Assume all memory is initialized to 0, and each of these two threads is spawned very close to the other in random order and execute in parallel.</p>
<p>What will each print?</p>
<pre tabindex="0"><code>x = 1
print( y )
</code></pre>
<pre tabindex="0"><code>y = 1
print( x )
</code></pre>
<p>Please do take some time to think this through.</p>
<p>The most natural answer is that either <span>Thread A</span>
or <span>Thread B</span>
starts first and runs to completion, with its twin executing in turn. This yields two possible outputs:</p>
<pre tabindex="0"><code>[0 1] ; Thread A executed first
[1 0] ; Thread B executed first
</code></pre>
<p>Where [A B] is the output for each thread, with one possibility per line.</p>
<p>If you’ve had your morning coffee, you’ve probably spotted a third possibility. The execution of both threads overlaps, and the first line from both <span>Thread A</span>
and <span>Thread B</span>
execute first, setting both <span>x = 1</span>
and <span>y = 1</span>
.</p>
<p>By the time the printing occurs, both will output <span>1</span>
. So there are three possible outcomes:</p>
<pre tabindex="0"><code>[0 1] ; Thread A executed first
[1 0] ; Thread B executed first
[1 1] ; Interleaved execution
</code></pre>
<p>If you haven’t seen a problem like this before and figured that out, that’s genuinely impressive.</p>
<p>Before you go patting yourself on the back too much though, there’s just one problem.</p>
<p><strong>This is not the correct answer.</strong></p>
<p>It relies on an implicit assumption of sequential consistency, where each thread executes its lines in order, and any modifications made are immediately and globally visible.</p>
<p>The correct answer depends on a question - what is your memory model?</p>
<p>Take the pseudocode above, translate it to the language of your choice that implements mutable shared memory. Then run it on whatever device you’re reading this on. You will occasionally get a fourth result:</p>
<pre tabindex="0"><code>[0 1] ; Thread A executed first
[1 0] ; Thread B executed first
[1 1] ; Interleaved execution
[0 0] ; ?!
</code></pre>
<h4>Memory Models</h4>
<p>Modern processors do not execute instructions in order, and will aggressively load values from memory before they’re needed.</p>
<p>There is a strong contract that this will never break single threaded execution. With shared memory concurrency, all bets are off.</p>
<p>In the case where both threads print <span>0</span>
, each thread has aggressively loaded the value it intends to print before executing the store, and the execution of both threads have also overlapped. The programs have been implicitly transformed into:</p>
<pre tabindex="0"><code>print( y )
x = 1
</code></pre>
<pre tabindex="0"><code>print( x )
y = 1
</code></pre>
<p>Intel provides a strong memory model with good guarantees, with a single caveat - “Loads may be reordered with older stores to different locations but not with older stores to the same location”.<sup><a href="#fn:1">1</a></sup> That is exactly what is happening here.</p>
<p>ARM’s memory model offers much less protection. Their equivalent document may alarm. <sup><a href="#fn:2">2</a></sup></p>
<h4>Temporal Fuzzing</h4>
<p>I hope I’ve convinced you that analyzing a four line program that touches shared memory is not trivial.</p>
<p>So what hope do we have building something larger? What happens when want to write programs that interact with more complicated concurrency models, such as databases, disks that can fail, or networks that can go down or partition?</p>
<p>What happens when you want to write a distributed database that doesn’t lose data, even when a disk fails during a network partition while rebalancing shards?</p>
<p>This is the motivation behind a new Clojure library I’m developing, tentatively called Temper.<sup><a href="#fn:3">3</a></sup> It simulates multiple threads of execution, reordering and interleaving operations in any valid way as defined by the pluggable concurrency model.</p>
<p>To circle back to our motivating example, we can implement two threads and execute it both under ARM and Intel memory models, as well under sequential consistency, which never reorders operations.</p>
<pre tabindex="0"><code>(defn thread-a []
  (mem/write-val :x 1)
  (mem/read-val :y))

(defn thread-b []
  (mem/write-val :y 1)
  (mem/read-val :x))

(st/multi-execute (Memory. {:model :intel}) {} [thread-a thread-b])
(st/multi-execute (Memory. {:model :arm}) {} [thread-a thread-b])
(st/multi-execute (Memory. {:model :sc}) {} [thread-a thread-b])
</code></pre>
<p>Each test is run 1000 times, and the frequency of each outcome is listed.</p>
<pre tabindex="0"><code>; Intel result
[0 1] ; 235 of 1000
[1 1] ; 268 of 1000
[1 0] ; 254 of 1000
[0 0] ; 243 of 1000
</code></pre>
<pre tabindex="0"><code>; ARM result
[0 0] ; 243 of 1000
[1 0] ; 230 of 1000
[1 1] ; 270 of 1000
[0 1] ; 257 of 1000
</code></pre>
<pre tabindex="0"><code>; Sequential Consistency result
[1 0] ; 242 of 1000
[1 1] ; 513 of 1000
[0 1] ; 245 of 1000
</code></pre>
<h4>Implementation</h4>
<p>Threads are standard Clojure code. Each read or write to the central data store immediately returns a promise. It also generates an event, which is added to a pending queue:</p>
<pre tabindex="0"><code>[0 :write :x 1]
[0 :read :y]
[1 :write :y 1]
[1 :read :x]
</code></pre>
<p>When all threads are parked and waiting for the result of a promise, the execution engine uses the active memory model to determine which events are legal to execute next. For instance, two writes on the same thread cannot have their order shuffled on Intel, but they can on ARM.</p>
<p>Executing an event modifies the global store, and returns the result of the operation to the promise. The thread blocking on that promise can now proceed.</p>
<p>This execution model flattens the probability distribution, and makes edge cases trigger often. In a real world test, <span>[0 0]</span>
is a result that may trigger a few times every million runs on an Intel chip. In the above simulation, its probability is 25%.</p>
<p>A new model can be implemented using this high level interface:</p>
<pre tabindex="0"><code>(defprotocol Model
  ; Updates store with the result of executing msg
  (apply-message [_ msg store])
  ; When fuzzing orders, can fst be reordered before lst?
  (order-valid? [_ fst lst])
  ; Can msg be executed given the state of store,
  ; used for locks and transactions
  (can-execute? [_ msg store]))
</code></pre>
<p>While so far we’ve focused on in process shared memory concurrency, the same technique can be used to simulate multiple processes accessing a disk, programs communicating over a network, database transactions at different isolation levels and more.</p>
<h4>Interactive Visualisation</h4>
<p>After execution, a Temper simulation can be dumped and visualised offline. The interleaved execution of each thread is shown, as well as the state of memory at any point in time. Each line of memory represents the state <em>after</em> the execution of the instruction to its left.</p>
<p>It’s important to note the reordering of instructions within each thread. Local reordering together with thread interleaving determine the program output.</p>
<p>Press the Shuffle button to fetch the next execution trace.</p>
<pre tabindex="0"><code>(defn thread-a []
  (mem/write-val :x 1)
  (mem/read-val :y))

(defn thread-b []
  (mem/write-val :y 1)
  (mem/read-val :x))
</code></pre>
<div data-res="
[{:results [0 1], :ops [{:thread 0, :order 1, :rec [:write :x 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 2, :rec [:read :x]} {:thread 1, :order 1, :rec [:write :y 1]}], :store-history [{:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 1}]} {:results [0 0], :ops [{:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 2, :rec [:read :y]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 1, :rec [:write :y 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 1}]} {:results [1 0], :ops [{:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 2, :rec [:read :y]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 1} {:x 1, :y 1}]} {:results [0 1], :ops [{:thread 0, :order 1, :rec [:write :x 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 1, :order 2, :rec [:read :x]}], :store-history [{:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 1} {:x 1, :y 1}]} {:results [0 1], :ops [{:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 1, :rec [:write :y 1]}], :store-history [{:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 1}]} {:results [1 0], :ops [{:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 1, :rec [:write :x 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 1} {:x 0, :y 1} {:x 0, :y 1} {:x 1, :y 1}]} {:results [1 0], :ops [{:thread 1, :order 1, :rec [:write :y 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 2, :rec [:read :y]} {:thread 0, :order 1, :rec [:write :x 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 1} {:x 0, :y 1} {:x 0, :y 1} {:x 1, :y 1}]} {:results [1 0], :ops [{:thread 1, :order 2, :rec [:read :x]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 0, :order 2, :rec [:read :y]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 1], :ops [{:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 2, :rec [:read :x]}], :store-history [{:x 0, :y 0} {:x 0, :y 1} {:x 1, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 1], :ops [{:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 2, :rec [:read :x]}], :store-history [{:x 0, :y 0} {:x 0, :y 1} {:x 0, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 0], :ops [{:thread 1, :order 1, :rec [:write :y 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 0, :order 2, :rec [:read :y]}], :store-history [{:x 0, :y 0} {:x 0, :y 1} {:x 0, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 0], :ops [{:thread 1, :order 2, :rec [:read :x]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 0, :order 1, :rec [:write :x 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 1} {:x 0, :y 1} {:x 1, :y 1}]} {:results [0 1], :ops [{:thread 0, :order 2, :rec [:read :y]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 1, :order 1, :rec [:write :y 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 1}]} {:results [0 0], :ops [{:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 1, :rec [:write :y 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 1}]} {:results [0 1], :ops [{:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 2, :rec [:read :x]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 1], :ops [{:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 2, :rec [:read :x]}], :store-history [{:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 1], :ops [{:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 2, :rec [:read :y]}], :store-history [{:x 0, :y 0} {:x 0, :y 1} {:x 1, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [0 0], :ops [{:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 2, :rec [:read :x]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 1, :rec [:write :x 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 1} {:x 1, :y 1}]} {:results [0 0], :ops [{:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 1, :rec [:write :x 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 1} {:x 1, :y 1}]} {:results [0 1], :ops [{:thread 0, :order 2, :rec [:read :y]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 1, :order 2, :rec [:read :x]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 1], :ops [{:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 2, :rec [:read :y]}], :store-history [{:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 1} {:x 1, :y 1} {:x 1, :y 1}]} {:results [1 1], :ops [{:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 0, :order 2, :rec [:read :y]}], :store-history [{:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 0} {:x 1, :y 1} {:x 1, :y 1}]} {:results [0 0], :ops [{:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 1, :rec [:write :x 1]} {:thread 1, :order 1, :rec [:write :y 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 0} {:x 1, :y 0} {:x 1, :y 1}]} {:results [0 0], :ops [{:thread 0, :order 2, :rec [:read :y]} {:thread 1, :order 1, :rec [:write :y 1]} {:thread 1, :order 2, :rec [:read :x]} {:thread 0, :order 1, :rec [:write :x 1]}], :store-history [{:x 0, :y 0} {:x 0, :y 0} {:x 0, :y 1} {:x 0, :y 1} {:x 1, :y 1}]}]

"><div><div><div>
<p><span>{<span>:x</span><span>0</span><span>:y</span><span>0</span>}</span></p>
<p><span>{<span>:x</span><span>1</span><span>:y</span><span>0</span>}</span></p>
<p><span>{<span>:x</span><span>1</span><span>:y</span><span>0</span>}</span></p>
<p><span>{<span>:x</span><span>1</span><span>:y</span><span>0</span>}</span></p>
<p><span>{<span>:x</span><span>1</span><span>:y</span><span>1</span>}</span></p>
</div></div></div></div>
<h4>Up Next</h4>
<p>In the <a href="https://www.reitzen.com/post/temporal-fuzzing-02/">next section</a>, we’re going to use this system to dive deeper into in process concurrency, including modeling locks and memory fences.</p>
<p>After that we can move on to the good stuff!</p></article></div><br>
<br>
<br>
<br>
<br>
<br>
via www.reitzen.com https://www.reitzen.com<br>
<br>
May 9, 2024 at 02:41PM
